src/operator/tensor/indexing_op.h

```diff
@@ -327,6 +328,51 @@ void TakeOpBackward(const nnvm::NodeAttrs& attrs,
     });
 }
 
+inline bool BatchTakeOpShape(const nnvm::NodeAttrs& attrs,
+                             std::vector<TShape> *in_attrs,
+                             std::vector<TShape> *out_attrs) {
+  CHECK_EQ(in_attrs->size(), 2) << "BatchTake op requires three inputs";
+  if ((*in_attrs)[1].ndim() != 0) {
+    SHAPE_ASSIGN_CHECK(*out_attrs, 0, (*in_attrs)[1]);
+  } else if ((*out_attrs)[0].ndim() != 0) {
+    SHAPE_ASSIGN_CHECK(*in_attrs, 1, (*out_attrs)[0]);
+  }
+  if ((*in_attrs)[0].ndim() == 0) return false;
+  CHECK_EQ((*in_attrs)[0].ndim(), 2) << "Data array must by 2 dimensional";
+  if ((*out_attrs)[0].ndim() == 0) return false;
+  CHECK_EQ((*in_attrs)[0][0], (*out_attrs)[0][0])
+    << "Index array's size must be the same as data array's first dimension";
+  return true;
+}
+
+inline bool BatchTakeOpType(const nnvm::NodeAttrs& attrs,
+                          std::vector<int> *in_attrs,
+                          std::vector<int> *out_attrs) {
+  CHECK_EQ(in_attrs->size(), 2);
+  if ((*in_attrs)[0] != -1) {
+    TYPE_ASSIGN_CHECK(*out_attrs, 0, (*in_attrs)[0]);
+  } else if ((*out_attrs)[0] != -1) {
+    TYPE_ASSIGN_CHECK(*in_attrs, 0, (*out_attrs)[0]);
+  }
+  TYPE_ASSIGN_CHECK(*in_attrs, 1, mshadow::kInt32);
```

Comment:
@piiswrong `batch_take` seems requires the indices to be an int array, while `take` is ok for float array.